detection_rules/devtools.py

```diff
@@ -1138,3 +1138,74 @@ def build_integration_manifests(overwrite: bool):
     integration_tags = list(set([r.contents.metadata.integration for r in rules if r.contents.metadata.integration]))
     click.echo(f"integration tags identified: {integration_tags}")
     build_integrations_manifest(overwrite, integration_tags)
+
+
+@dev_group.group('attack')
+def attack_group():
+    """Commands for managing Mitre ATT&CK data and mappings."""
+
+
+@attack_group.command('refresh-data')
+def refresh_attack_data() -> dict:
+    """Refresh the ATT&CK data file."""
+    data, _ = attack.refresh_attack_data()
+    return data
+
+
+@attack_group.command('refresh-redirect-mappings')
+def refresh_threat_mappings():
+    """Refresh the ATT&CK redirect file and update all rule threat mappings."""
+    # refresh the attack_technique_redirects
+    click.echo('refreshing data in attack_technique_redirects.json')
+    attack.refresh_redirected_techniques_map()
```

Comment:
Why do we follow the redirection if we have an updated copy of the entire matrix from the `refresh-data` command? We might want to explore all of the given data in a bundle locally.

I downloaded the 11.3 bundle, loaded the `enterprise-attack.json` file and it contains all attack objects, including revoked and deprecated values and flags that could be referenced instead. This could reduce the number of network calls needed for this.

<details>
<summary>Revoked object example with source and target references</summary>

```
{'object_marking_refs': ['marking-definition--fa42a846-8d90-4e51-bc29-71d5b4802168'],
 'id': 'relationship--632ca9a0-a9f3-4b27-96e1-9fcb8bab11cb',
 'type': 'relationship',
 'created': '2018-10-17T00:14:20.652Z',
 'created_by_ref': 'identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5',
 'modified': '2018-10-17T00:14:20.652Z',
 'relationship_type': 'revoked-by',
 'source_ref': 'intrusion-set--68ba94ab-78b8-43e7-83e2-aed3466882c6',
 'target_ref': 'intrusion-set--4ca1929c-7d64-4aab-b849-badbfc0c760d',
 'x_mitre_version': '1.0',
 'x_mitre_modified_by_ref': 'identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5'}
```

</details>

<details>
<summary>Deprecated object example</summary>

```
{'x_mitre_domains': ['enterprise-attack'],
 'object_marking_refs': ['marking-definition--fa42a846-8d90-4e51-bc29-71d5b4802168'],
 'id': 'course-of-action--03c0c586-50ed-45a7-95f4-f496d7eb5330',
 'type': 'course-of-action',
 'created': '2018-10-17T00:14:20.652Z',
 'created_by_ref': 'identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5',
 'external_references': [{'source_name': 'mitre-attack',
   'url': 'https://attack.mitre.org/mitigations/T1148',
   'external_id': 'T1148'},
  {'url': 'http://www.akyl.net/securing-bashhistory-file-make-sure-your-linux-system-users-won%E2%80%99t-hide-or-delete-their-bashhistory',
   'description': 'Mathew Branwell. (2012, March 21). Securing .bash_history file. Retrieved July 8, 2017.',
   'source_name': 'Securing bash history'}],
 'modified': '2019-07-24T19:34:34.065Z',
 'name': 'HISTCONTROL Mitigation',
 'description': 'Prevent users from changing the <code>HISTCONTROL</code> environment variable (Citation: Securing bash history). Also, make sure that the <code>HISTCONTROL</code> environment variable is set to “ignoredup” instead of “ignoreboth” or “ignorespace”.',
 'x_mitre_deprecated': True,
 'x_mitre_version': '1.0',
 'x_mitre_modified_by_ref': 'identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5'}
```

</details>

Reply:
good point, I will look into building the redirect from the attack deprecations directly instead

Reply:
Here is an example of an old subtechnique being moved to a new technique (`T1547.011` to `T1647`). How is the relationship conveyed in the data that it is superseding it?

<details>
<summary>Old: T1547.011</summary><br>

```json
        {
            "created": "2020-01-24T20:02:59.149Z",
            "created_by_ref": "identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5",
            "description": "Adversaries can modify property list files (plist files) to execute their code as part of establishing persistence. Plist files are used by macOS applications to store properties and configuration settings for applications and services. Applications use information plist files, <code>Info.plist</code>, to tell the operating system how to handle the application at runtime using structured metadata in the form of keys and values. Plist files are formatted in XML and based on Apple's Core Foundation DTD and can be saved in text or binary format.(Citation: fileinfo plist file description) \n\nAdversaries can modify paths to executed binaries, add command line arguments, and insert key/pair values to plist files in auto-run locations which execute upon user logon or system startup. Through modifying plist files in these locations, adversaries can also execute a malicious dynamic library (dylib) by adding a dictionary containing the <code>DYLD_INSERT_LIBRARIES</code> key combined with a path to a malicious dylib under the <code>EnvironmentVariables</code> key in a plist file. Upon user logon, the plist is called for execution and the malicious dylib is executed within the process space. Persistence can also be achieved by modifying the <code>LSEnvironment</code> key in the application's <code>Info.plist</code> file.(Citation: wardle artofmalware volume1)",
            "external_references": [
                {
                    "external_id": "T1547.011",
                    "source_name": "mitre-attack",
                    "url": "https://attack.mitre.org/techniques/T1547/011"
                },
                {
                    "description": "FileInfo.com team. (2019, November 26). .PLIST File Extension. Retrieved October 12, 2021.",
                    "source_name": "fileinfo plist file description",
                    "url": "https://fileinfo.com/extension/plist"
                },
                {
                    "description": "Patrick Wardle. (2020, August 5). The Art of Mac Malware Volume 0x1: Analysis. Retrieved March 19, 2021.",
                    "source_name": "wardle artofmalware volume1",
                    "url": "https://taomm.org/vol1/pdfs.html"
                }
            ],
            "id": "attack-pattern--6747daa2-3533-4e78-8fb8-446ebb86448a",
            "kill_chain_phases": [
                {
                    "kill_chain_name": "mitre-attack",
                    "phase_name": "persistence"
                },
                {
                    "kill_chain_name": "mitre-attack",
                    "phase_name": "privilege-escalation"
                }
            ],
            "modified": "2022-04-20T21:06:07.560Z",
            "name": "Plist Modification",
            "object_marking_refs": [
                "marking-definition--fa42a846-8d90-4e51-bc29-71d5b4802168"
            ],
            "revoked": true,
            "type": "attack-pattern",
            "x_mitre_attack_spec_version": "2.1.0",
            "x_mitre_data_sources": [
                "File: File Modification",
                "Service: Service Creation",
                "Command: Command Execution",
                "Process: Process Creation"
            ],
            "x_mitre_deprecated": false,
            "x_mitre_detection": "Monitor for common command-line editors used to modify plist files located in auto-run locations, such as <code>~/LaunchAgents</code>, <code>~/Library/Application Support/com.apple.backgroundtaskmanagementagent/backgrounditems.btm</code>, and an application's <code>Info.plist</code>. \n\nMonitor for plist file modification immediately followed by code execution from <code>~/Library/Scripts</code> and <code>~/Library/Preferences</code>. Also, monitor for significant changes to any path pointers in a modified plist.\n\nIdentify new services executed from plist modified in the previous user's session. ",
            "x_mitre_domains": [
                "enterprise-attack"
            ],
            "x_mitre_is_subtechnique": true,
            "x_mitre_modified_by_ref": "identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5",
            "x_mitre_permissions_required": [
                "User",
                "Administrator"
            ],
            "x_mitre_platforms": [
                "macOS"
            ],
            "x_mitre_version": "1.1"
        }
```

</details>


<details>
<summary>New: T1647</summary><br>

```json
        {
            "created": "2022-04-09T15:06:32.458Z",
            "created_by_ref": "identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5",
            "description": "Adversaries may modify property list files (plist files) to enable other malicious activity, while also potentially evading and bypassing system defenses. macOS applications use plist files, such as the <code>info.plist</code> file, to store properties and configuration settings that inform the operating system how to handle the application at runtime. Plist files are structured metadata in key-value pairs formatted in XML based on Apple's Core Foundation DTD. Plist files can be saved in text or binary format.(Citation: fileinfo plist file description) \n\nAdversaries can modify key-value pairs in plist files to influence system behaviors, such as hiding the execution of an application (i.e. [Hidden Window](https://attack.mitre.org/techniques/T1564/003)) or running additional commands for persistence (ex: [Launch Agent](https://attack.mitre.org/techniques/T1543/001)/[Launch Daemon](https://attack.mitre.org/techniques/T1543/004) or [Re-opened Applications](https://attack.mitre.org/techniques/T1547/007)).\n\nFor example, adversaries can add a malicious application path to the `~/Library/Preferences/com.apple.dock.plist` file, which controls apps that appear in the Dock. Adversaries can also modify the <code>LSUIElement</code> key in an application\u2019s <code>info.plist</code> file  to run the app in the background. Adversaries can also insert key-value pairs to insert environment variables, such as <code>LSEnvironment</code>, to enable persistence via [Dynamic Linker Hijacking](https://attack.mitre.org/techniques/T1574/006).(Citation: wardle chp2 persistence)(Citation: eset_osx_flashback)",
            "external_references": [
                {
                    "external_id": "T1647",
                    "source_name": "mitre-attack",
                    "url": "https://attack.mitre.org/techniques/T1647"
                },
                {
                    "description": "ESET. (2012, January 1). OSX/Flashback. Retrieved April 19, 2022.",
                    "source_name": "eset_osx_flashback",
                    "url": "https://www.welivesecurity.com/wp-content/uploads/200x/white-papers/osx_flashback.pdf"
                },
                {
                    "description": "FileInfo.com team. (2019, November 26). .PLIST File Extension. Retrieved October 12, 2021.",
                    "source_name": "fileinfo plist file description",
                    "url": "https://fileinfo.com/extension/plist"
                },
                {
                    "description": "Patrick Wardle. (2022, January 1). The Art of Mac Malware Volume 0x1:Analysis. Retrieved April 19, 2022.",
                    "source_name": "wardle chp2 persistence",
                    "url": "https://taomm.org/PDFs/vol1/CH%200x02%20Persistence.pdf"
                }
            ],
            "id": "attack-pattern--7d20fff9-8751-404e-badd-ccd71bda0236",
            "kill_chain_phases": [
                {
                    "kill_chain_name": "mitre-attack",
                    "phase_name": "defense-evasion"
                }
            ],
            "modified": "2022-04-20T22:00:33.375Z",
            "name": "Plist File Modification",
            "object_marking_refs": [
                "marking-definition--fa42a846-8d90-4e51-bc29-71d5b4802168"
            ],
            "revoked": false,
            "type": "attack-pattern",
            "x_mitre_attack_spec_version": "2.1.0",
            "x_mitre_data_sources": [
                "Service: Service Creation",
                "Command: Command Execution",
                "Process: Process Creation",
                "File: File Modification"
            ],
            "x_mitre_deprecated": false,
            "x_mitre_detection": "Monitor for common command-line editors used to modify plist files located in auto-run locations, such as <code>\\~/LaunchAgents</code>, <code>~/Library/Application Support/com.apple.backgroundtaskmanagementagent/backgrounditems.btm</code>, and an application's <code>Info.plist</code>. \n\nMonitor for plist file modification immediately followed by code execution from <code>\\~/Library/Scripts</code> and <code>~/Library/Preferences</code>. Also, monitor for significant changes to any path pointers in a modified plist.\n\nIdentify new services executed from plist modified in the previous user's session. ",
            "x_mitre_domains": [
                "enterprise-attack"
            ],
            "x_mitre_is_subtechnique": false,
            "x_mitre_modified_by_ref": "identity--c78cb6e5-0c4b-4611-8297-d1b8b55e40b5",
            "x_mitre_platforms": [
                "macOS"
            ],
            "x_mitre_version": "1.0"
        }
```

</details>